File: 968.监控二叉树.java
/*
 * @lc app=leetcode.cn id=968 lang=java
 *
 * [968] 监控二叉树
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static enum State {
        UNCOVERED,
        COVERED_WO_CAM,
        COVERED_W_CAM
    }

    private int ans;

    public int minCameraCover(TreeNode root) {
        ans = 0;
        if (postOrder(root) == State.UNCOVERED) {
            // Root's children are covered without a camera,
            // but the root itself is uncovered 
            return ans + 1;
        }
        return ans;
    }

    private State postOrder(TreeNode node) {
        if (node == null)  return State.COVERED_WO_CAM;
        State leftState = postOrder(node.left);
        State rightState = postOrder(node.right);
        if (leftState == State.UNCOVERED || rightState == State.UNCOVERED) {
            ++ans;
            return State.COVERED_W_CAM;
        }
        if (leftState == State.COVERED_W_CAM || rightState == State.COVERED_W_CAM) {
            return State.COVERED_WO_CAM;
        }
        // States of children both equal State.COVERED_WO_CAM
        return State.UNCOVERED;
    }
}
// @lc code=end

File: 968.监控二叉树.py
#
# @lc app=leetcode.cn id=968 lang=python3
#
# [968] 监控二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minCameraCover(self, root: TreeNode) -> int:
        """Traversal:
            Post-order is appropriate because we are finding
            the minimum number of camera so the first few 
            cameras should be placed at parent of leaf nodes.
        
        Cases on node:
            0. not covered
            1. has camera
            2. covered
        """
        ans = 0

        def traverse(node):
            if not node:
                # null nodes are assumed covered
                return 2
            nonlocal ans

            '''Post-order traversal'''
            left = traverse(node.left)
            right = traverse(node.right)
            if left == 2 and right == 2:
                # both children are covered, current node not covered
                return 0
            elif not left or not right:
                # current node has or needs a camera
                ans += 1
                return 1
            else:    # if left == 1 or right == 1:
                # current node covered
                return 2
        
        if traverse(root) == 0:
            # root is not yet covered
            ans += 1
        return ans
        
# @lc code=end

